File: src/qubitcoin/privacy/stealth.py
"""
Stealth Address System for Qubitcoin Susy Swaps

Stealth addresses provide recipient privacy by generating a unique,
one-time address for every transaction. Even if an observer knows
the recipient's public address, they cannot link received transactions.

Protocol:
  1. Recipient publishes (spend_pubkey, view_pubkey).
  2. Sender generates ephemeral keypair (r, R = r*G).
  3. Sender computes shared_secret = hash(r * view_pubkey).
  4. One-time address = spend_pubkey + shared_secret * G.
  5. Sender publishes R (ephemeral pubkey) in the transaction.
  6. Recipient scans: for each R, computes shared_secret = hash(view_privkey * R),
     checks if spend_pubkey + shared_secret*G matches the output address.
  7. If match, recipient can spend using spend_privkey + shared_secret.
"""
import hashlib
import os
from typing import Tuple, Optional
from dataclasses import dataclass

from .commitments import (
    ECPoint, G, INFINITY, _N, _P,
    _scalar_mult, _point_add,
)
from ..utils.logger import get_logger

logger = get_logger(__name__)


@dataclass
class StealthKeyPair:
    """A stealth address key pair (spend + view)."""
    spend_privkey: int
    spend_pubkey: ECPoint
    view_privkey: int
    view_pubkey: ECPoint

    def public_address(self) -> str:
        """Get the publishable stealth address (spend_pub || view_pub)."""
        def compress(p: ECPoint) -> bytes:
            prefix = b'\x02' if p.y % 2 == 0 else b'\x03'
            return prefix + p.x.to_bytes(32, 'big')
        return (compress(self.spend_pubkey) + compress(self.view_pubkey)).hex()


@dataclass
class StealthOutput:
    """A stealth transaction output."""
    one_time_address: ECPoint  # The one-time destination address
    ephemeral_pubkey: ECPoint  # R = r*G, published in transaction
    shared_secret: int          # Used internally, not published

    def address_hex(self) -> str:
        """Get hex representation of one-time address."""
        p = self.one_time_address
        prefix = b'\x02' if p.y % 2 == 0 else b'\x03'
        return (prefix + p.x.to_bytes(32, 'big')).hex()

    def ephemeral_hex(self) -> str:
        """Get hex representation of ephemeral pubkey."""
        p = self.ephemeral_pubkey
        prefix = b'\x02' if p.y % 2 == 0 else b'\x03'
        return (prefix + p.x.to_bytes(32, 'big')).hex()


def _hash_to_scalar(point: ECPoint) -> int:
    """Hash an EC point to a scalar (Fiat-Shamir style)."""
    if point.is_infinity:
        data = b'\x00' * 33
    else:
        prefix = b'\x02' if point.y % 2 == 0 else b'\x03'
        data = prefix + point.x.to_bytes(32, 'big')
    h = hashlib.sha256(b"Qubitcoin_stealth_v1" + data).digest()
    return int.from_bytes(h, 'big') % _N


class StealthAddressManager:
    """Manage stealth addresses for privacy transactions."""

    @staticmethod
    def generate_keypair() -> StealthKeyPair:
        """Generate a new stealth key pair (spend + view keys).

        Returns:
            StealthKeyPair with both spend and view private/public keys.
        """
        spend_priv = int.from_bytes(os.urandom(32), 'big') % _N
        view_priv = int.from_bytes(os.urandom(32), 'big') % _N
        spend_pub = _scalar_mult(spend_priv, G)
        view_pub = _scalar_mult(view_priv, G)
        return StealthKeyPair(
            spend_privkey=spend_priv,
            spend_pubkey=spend_pub,
            view_privkey=view_priv,
            view_pubkey=view_pub,
        )

    @staticmethod
    def create_output(recipient_spend_pub: ECPoint,
                      recipient_view_pub: ECPoint) -> StealthOutput:
        """Create a stealth output for a recipient (sender side).

        Args:
            recipient_spend_pub: Recipient's public spend key.
            recipient_view_pub: Recipient's public view key.

        Returns:
            StealthOutput containing the one-time address and ephemeral key.
        """
        # Generate ephemeral keypair
        r = int.from_bytes(os.urandom(32), 'big') % _N
        R = _scalar_mult(r, G)

        # Shared secret = hash(r * view_pub)
        shared_point = _scalar_mult(r, recipient_view_pub)
        shared_secret = _hash_to_scalar(shared_point)

        # One-time address = spend_pub + shared_secret * G
        ss_G = _scalar_mult(shared_secret, G)
        one_time_addr = _point_add(recipient_spend_pub, ss_G)

        return StealthOutput(
            one_time_address=one_time_addr,
            ephemeral_pubkey=R,
            shared_secret=shared_secret,
        )

    @staticmethod
    def scan_output(keypair: StealthKeyPair, ephemeral_pubkey: ECPoint,
                    output_address: ECPoint) -> bool:
        """Scan a transaction output to check if it belongs to us (recipient side).

        Args:
            keypair: Our stealth key pair.
            ephemeral_pubkey: R from the transaction.
            output_address: The one-time address in the output.

        Returns:
            True if this output is destined for us.
        """
        # Compute shared secret = hash(view_priv * R)
        shared_point = _scalar_mult(keypair.view_privkey, ephemeral_pubkey)
        shared_secret = _hash_to_scalar(shared_point)

        # Expected address = spend_pub + shared_secret * G
        ss_G = _scalar_mult(shared_secret, G)
        expected = _point_add(keypair.spend_pubkey, ss_G)

        return expected == output_address

    @staticmethod
    def compute_spending_key(keypair: StealthKeyPair,
                             ephemeral_pubkey: ECPoint) -> int:
        """Compute the private key to spend a stealth output (recipient side).

        Args:
            keypair: Our stealth key pair.
            ephemeral_pubkey: R from the transaction we want to spend.

        Returns:
            The private key that controls the one-time address.
        """
        # shared_secret = hash(view_priv * R)
        shared_point = _scalar_mult(keypair.view_privkey, ephemeral_pubkey)
        shared_secret = _hash_to_scalar(shared_point)

        # spending_key = spend_priv + shared_secret (mod N)
        return (keypair.spend_privkey + shared_secret) % _N

    @staticmethod
    def compute_key_image(spending_key: int) -> ECPoint:
        """Compute a key image for double-spend prevention.

        Key image I = spending_key * hash_to_point(spending_key * G)
        The key image is deterministic per spending key, so attempting to
        spend the same output twice produces the same key image.

        Args:
            spending_key: The private spending key.

        Returns:
            Key image point.
        """
        pub = _scalar_mult(spending_key, G)
        # hash_to_point: we hash the public key and multiply
        hp = _hash_to_scalar(pub)
        hp_point = _scalar_mult(hp, G)
        return _scalar_mult(spending_key, hp_point)

File: aether-cli/crates/aether-privacy/src/stealth.rs
//! Stealth Address Protocol for Qubitcoin.
//!
//! Uses Diffie-Hellman on secp256k1 to generate one-time addresses:
//!
//! Recipient publishes: (spend_pub, view_pub)
//! Sender:
//!   1. Generates ephemeral keypair (r, R = r*G)
//!   2. Computes shared_secret = SHA256(r * view_pub)
//!   3. Derives one_time_address = spend_pub + shared_secret * G
//!   4. Publishes R in transaction
//! Recipient:
//!   1. Computes shared_secret = SHA256(view_priv * R)
//!   2. Checks if spend_pub + shared_secret * G == output address
//!   3. If match, spending_key = spend_priv + shared_secret
//!
//! Key images prevent double-spending: I = spend_key * H(spend_key * G)

use k256::elliptic_curve::group::GroupEncoding;
use k256::elliptic_curve::ops::Reduce;
use k256::elliptic_curve::sec1::ToEncodedPoint;
use k256::{AffinePoint, ProjectivePoint, Scalar, U256};
use rand::RngCore;
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};

/// Stealth address keypair (spend + view).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StealthKeypair {
    pub spend_privkey: Vec<u8>,
    pub spend_pubkey: Vec<u8>,
    pub view_privkey: Vec<u8>,
    pub view_pubkey: Vec<u8>,
}

impl StealthKeypair {
    /// The stealth public address: spend_pubkey || view_pubkey (66 bytes compressed).
    pub fn public_address(&self) -> Vec<u8> {
        let mut addr = Vec::with_capacity(66);
        addr.extend_from_slice(&self.spend_pubkey);
        addr.extend_from_slice(&self.view_pubkey);
        addr
    }

    /// Hex-encoded public address (132 chars).
    pub fn public_address_hex(&self) -> String {
        hex::encode(self.public_address())
    }
}

/// A stealth output created by a sender.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StealthOutput {
    /// One-time address (compressed point, 33 bytes).
    pub one_time_address: Vec<u8>,
    /// Sender's ephemeral public key R (compressed, 33 bytes).
    pub ephemeral_pubkey: Vec<u8>,
}

/// Key image for double-spend prevention.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct KeyImage {
    pub image: Vec<u8>,
}

pub struct StealthAddressManager;

impl StealthAddressManager {
    /// Generate a new stealth keypair (spend + view).
    pub fn generate_keypair() -> StealthKeypair {
        let (spend_priv, spend_pub) = generate_keypair();
        let (view_priv, view_pub) = generate_keypair();

        StealthKeypair {
            spend_privkey: spend_priv.to_bytes().to_vec(),
            spend_pubkey: spend_pub.to_affine().to_bytes().to_vec(),
            view_privkey: view_priv.to_bytes().to_vec(),
            view_pubkey: view_pub.to_affine().to_bytes().to_vec(),
        }
    }

    /// Create a stealth output for a recipient.
    /// Takes the recipient's spend_pub and view_pub (33 bytes each, compressed).
    pub fn create_output(
        recipient_spend_pub: &[u8],
        recipient_view_pub: &[u8],
    ) -> Result<StealthOutput, String> {
        let spend_pub = decode_point(recipient_spend_pub)?;
        let view_pub = decode_point(recipient_view_pub)?;

        // Generate ephemeral keypair
        let (eph_priv, eph_pub) = generate_keypair();

        // shared_secret = SHA256(r * view_pub)
        let shared_point = view_pub * eph_priv;
        let shared_secret = hash_shared_point(&shared_point);

        // one_time_address = spend_pub + shared_secret * G
        let ss_scalar = <Scalar as Reduce<U256>>::reduce_bytes(&shared_secret.into());
        let one_time = spend_pub + ProjectivePoint::GENERATOR * ss_scalar;

        Ok(StealthOutput {
            one_time_address: one_time.to_affine().to_bytes().to_vec(),
            ephemeral_pubkey: eph_pub.to_affine().to_bytes().to_vec(),
        })
    }

    /// Scan a stealth output to check if it belongs to us.
    /// Returns the spending key scalar if it matches.
    pub fn scan_output(
        keypair: &StealthKeypair,
        ephemeral_pubkey: &[u8],
        output_address: &[u8],
    ) -> Result<Option<Vec<u8>>, String> {
        let eph_pub = decode_point(ephemeral_pubkey)?;
        let view_priv = decode_scalar(&keypair.view_privkey)?;
        let spend_priv = decode_scalar(&keypair.spend_privkey)?;
        let spend_pub = decode_point(&keypair.spend_pubkey)?;

        // shared_secret = SHA256(view_priv * R)
        let shared_point = eph_pub * view_priv;
        let shared_secret = hash_shared_point(&shared_point);
        let ss_scalar = <Scalar as Reduce<U256>>::reduce_bytes(&shared_secret.into());

        // Expected one_time_address = spend_pub + shared_secret * G
        let expected = spend_pub + ProjectivePoint::GENERATOR * ss_scalar;
        let expected_bytes = expected.to_affine().to_bytes();

        if expected_bytes.as_slice() == output_address {
            // Derive spending key: spend_priv + shared_secret
            let spending_key = spend_priv + ss_scalar;
            Ok(Some(spending_key.to_bytes().to_vec()))
        } else {
            Ok(None)
        }
    }

    /// Compute key image for double-spend prevention.
    /// I = spending_key * H(spending_key * G)
    pub fn compute_key_image(spending_key: &[u8]) -> Result<KeyImage, String> {
        let sk = decode_scalar(spending_key)?;
        let pk = ProjectivePoint::GENERATOR * sk;

        // H(pk) - hash the public key to get a curve point
        let pk_bytes = pk.to_affine().to_bytes();
        let hp = hash_to_point(&pk_bytes);

        // I = sk * H(pk)
        let image = hp * sk;

        Ok(KeyImage {
            image: image.to_affine().to_bytes().to_vec(),
        })
    }
}

// ── Helpers ──────────────────────────────────────────────────────

fn generate_keypair() -> (Scalar, ProjectivePoint) {
    let mut bytes = [0u8; 32];
    rand::thread_rng().fill_bytes(&mut bytes);
    let scalar = <Scalar as Reduce<U256>>::reduce_bytes(&bytes.into());
    let point = ProjectivePoint::GENERATOR * scalar;
    (scalar, point)
}

fn decode_point(bytes: &[u8]) -> Result<ProjectivePoint, String> {
    let arr: [u8; 33] = bytes
        .try_into()
        .map_err(|_| format!("point must be 33 bytes compressed, got {}", bytes.len()))?;
    let affine = AffinePoint::from_bytes(&arr.into());
    if affine.is_some().into() {
        Ok(ProjectivePoint::from(affine.unwrap()))
    } else {
        Err("invalid curve point".to_string())
    }
}

fn decode_scalar(bytes: &[u8]) -> Result<Scalar, String> {
    if bytes.len() != 32 {
        return Err(format!("scalar must be 32 bytes, got {}", bytes.len()));
    }
    let mut arr = [0u8; 32];
    arr.copy_from_slice(bytes);
    Ok(<Scalar as Reduce<U256>>::reduce_bytes(&arr.into()))
}

fn hash_shared_point(point: &ProjectivePoint) -> [u8; 32] {
    let encoded = point.to_affine().to_encoded_point(false);
    let mut hasher = Sha256::new();
    hasher.update(b"QBC_stealth_shared_secret_v1");
    hasher.update(encoded.as_bytes());
    hasher.finalize().into()
}

/// Hash arbitrary data to a curve point (try-and-increment).
fn hash_to_point(data: &[u8]) -> ProjectivePoint {
    let mut counter = 0u32;
    loop {
        let mut hasher = Sha256::new();
        hasher.update(b"QBC_hash_to_point_v1");
        hasher.update(data);
        hasher.update(counter.to_le_bytes());
        let hash: [u8; 32] = hasher.finalize().into();

        let mut compressed = [0u8; 33];
        compressed[0] = 0x02;
        compressed[1..].copy_from_slice(&hash);

        let point = AffinePoint::from_bytes(&compressed.into());
        if point.is_some().into() {
            return ProjectivePoint::from(point.unwrap());
        }
        counter += 1;
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_generate_keypair() {
        let kp = StealthAddressManager::generate_keypair();
        assert_eq!(kp.spend_privkey.len(), 32);
        assert_eq!(kp.spend_pubkey.len(), 33);
        assert_eq!(kp.view_privkey.len(), 32);
        assert_eq!(kp.view_pubkey.len(), 33);
        assert_eq!(kp.public_address().len(), 66);
    }

    #[test]
    fn test_create_and_scan_output() {
        let recipient = StealthAddressManager::generate_keypair();
        let output = StealthAddressManager::create_output(
            &recipient.spend_pubkey,
            &recipient.view_pubkey,
        )
        .unwrap();

        assert_eq!(output.one_time_address.len(), 33);
        assert_eq!(output.ephemeral_pubkey.len(), 33);

        // Recipient scans and finds it
        let spending_key = StealthAddressManager::scan_output(
            &recipient,
            &output.ephemeral_pubkey,
            &output.one_time_address,
        )
        .unwrap();

        assert!(spending_key.is_some());
        assert_eq!(spending_key.unwrap().len(), 32);
    }

    #[test]
    fn test_scan_wrong_recipient_returns_none() {
        let recipient = StealthAddressManager::generate_keypair();
        let other = StealthAddressManager::generate_keypair();

        let output = StealthAddressManager::create_output(
            &recipient.spend_pubkey,
            &recipient.view_pubkey,
        )
        .unwrap();

        // Other person scans — should NOT match
        let result = StealthAddressManager::scan_output(
            &other,
            &output.ephemeral_pubkey,
            &output.one_time_address,
        )
        .unwrap();

        assert!(result.is_none());
    }

    #[test]
    fn test_key_image_deterministic() {
        let kp = StealthAddressManager::generate_keypair();
        let output = StealthAddressManager::create_output(
            &kp.spend_pubkey,
            &kp.view_pubkey,
        )
        .unwrap();

        let sk = StealthAddressManager::scan_output(
            &kp,
            &output.ephemeral_pubkey,
            &output.one_time_address,
        )
        .unwrap()
        .unwrap();

        let ki1 = StealthAddressManager::compute_key_image(&sk).unwrap();
        let ki2 = StealthAddressManager::compute_key_image(&sk).unwrap();
        assert_eq!(ki1.image, ki2.image);
    }

    #[test]
    fn test_different_outputs_different_addresses() {
        let kp = StealthAddressManager::generate_keypair();
        let o1 = StealthAddressManager::create_output(&kp.spend_pubkey, &kp.view_pubkey).unwrap();
        let o2 = StealthAddressManager::create_output(&kp.spend_pubkey, &kp.view_pubkey).unwrap();
        // Different ephemeral keys → different one-time addresses
        assert_ne!(o1.one_time_address, o2.one_time_address);
        assert_ne!(o1.ephemeral_pubkey, o2.ephemeral_pubkey);
    }

    #[test]
    fn test_public_address_hex() {
        let kp = StealthAddressManager::generate_keypair();
        let hex_addr = kp.public_address_hex();
        assert_eq!(hex_addr.len(), 132); // 66 bytes * 2
    }
}
